### ISSUE TITLE
Fragment scope not automatically linked to activity scope

### ISSUE THREAD

**Issue by bmesing** [InsertKoinIO/koin#1087]
Contrary to what the documentation suggests, fragment scopes are not linked to activity scopes automatically. This is because fragmentScope() LifecycleScopeDelegate is constructed upon fragment instance creation, when fragment is not yet attached to activity.

ScopeFragment defines:
`  override val scope: Scope by fragmentScope()
`

with 
```
fun Fragment.fragmentScope() = LifecycleScopeDelegate(this){ koin: Koin ->
    val scope = koin.createScope(getScopeId(), getScopeName())
    val activityScope = activity?.getScopeOrNull()
    activityScope?.let { scope.linkTo(it) }
    scope
}
```

and LIfecycleScopeDelegate

```
class LifecycleScopeDelegate(
        val lifecycleOwner: LifecycleOwner,
        koinContext: KoinContext = GlobalContext,
        createScope: (Koin) -> Scope = { koin: Koin -> koin.createScope(lifecycleOwner.getScopeId(), lifecycleOwner.getScopeName()) },
) : ReadOnlyProperty<LifecycleOwner, Scope> {

    private var _scope: Scope? = null

    init {
        val koin = koinContext.get()
        val logger = koin.logger

        logger.debug("setup scope: $_scope for $lifecycleOwner")
        val scopeId = lifecycleOwner.getScopeId()
        // here the scope get created while the fragment is not yet attached
        _scope = koin.getScopeOrNull(scopeId) ?: createScope(koin)
        logger.debug("got scope: $_scope for $lifecycleOwner")

        lifecycleOwner.lifecycle.addObserver(object : LifecycleObserver {
            @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
            fun onDestroy(owner: LifecycleOwner) {
                logger.debug("Closing scope: $_scope for $lifecycleOwner")
                if (_scope?.closed == false) {
                    _scope?.close()
                }
                _scope = null
            }
        })
    }
    override fun getValue(thisRef: LifecycleOwner, property: KProperty<*>): Scope {
        return _scope ?: error("can't get Scope for $lifecycleOwner")
    }
}
```

**Comment by bmesing** [comment]
@arnaudgiuliani note, that his is not a documentation issue, but a bug, because they are supposed to be linked together. This is something quite central for koin Android and I feel this should be adressed.

**Comment by arnaudgiuliani** [comment]
Are you sure it's not linked?

in the block above, we are linking it:

```kotlin
fun Fragment.fragmentScope() = LifecycleScopeDelegate(this){ koin: Koin ->
    val scope = koin.createScope(getScopeId(), getScopeName())
    val activityScope = activity?.getScopeOrNull()
    activityScope?.let { scope.linkTo(it) }
    scope
}

```

**Comment by bmesing** [comment]
Yes I am:
"This is because fragmentScope() LifecycleScopeDelegate is constructed upon fragment instance creation, when fragment is not yet attached to activity."
So `activity?` in the third line always returns null.

Our local workaround is to manually add in our custom BaseFragment
```
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // need to link scope manually because of https://github.com/InsertKoinIO/koin/issues/1087
        val activityScope = activity?.getScopeOrNull()
        activityScope?.let { scope.linkTo(it) }
    }
```

**Comment by arnaudgiuliani** [comment]
Ok interesting 🤔 Let me check on a sample 👍

**Comment by arnaudgiuliani** [comment]
The code snippet you are mentionning is outdated. In latest 3.1.5, source is following:

```kotlin
lifecycleOwner.lifecycle.addObserver(object : DefaultLifecycleObserver {
            override fun onCreate(owner: LifecycleOwner) {
                createScope()
            }

            override fun onDestroy(owner: LifecycleOwner) {
                logger.debug("Closing scope: $_scope for $lifecycleOwner")
                if (_scope?.closed == false) {
                    _scope?.close()
                }
                _scope = null
            }
        })
```

The scope creation really occurs when we create the component with `onCreate`

**Comment by arnaudgiuliani** [comment]
I believe this has been fixed since a moment.

**Comment by 18500374995** [comment]
When I use koinApplication to replace startKoin  in library, There is one expection that `KoinApplication has not been started`.

```
override val scope: Scope by fragmentScope()
```
with
```
fun Fragment.fragmentScope() = LifecycleScopeDelegate(this){ koin: Koin ->
    val scope = koin.createScope(getScopeId(), getScopeName())
    val activityScope = activity?.getScopeOrNull()
    activityScope?.let { scope.linkTo(it) }
    scope
}
```
And I used third activity, the activity didn't implement `ScopeActivity` or `KoinCompent`, so it can't get koin instance.

```
fun ComponentActivity.getScopeOrNull(): Scope? = getKoin().getScopeOrNull(getScopeId())
```
with
```
fun ComponentCallbacks.getKoin() = when (this) {
    is KoinComponent -> this.getKoin()
    else -> GlobalContext.get()
}
```
with
```
    override fun get(): Koin = _koin ?: error("KoinApplication has not been started")
```

**Comment by bmesing** [comment]
@arnaudgiuliani I can confirm that this is fixed with 3.1.5.
Thanks for the fix and the great work on koin! 👍
